File: ruby/robot-name/robot_name.rb
class Robot
  attr_reader :name

  def initialize = reset
  def reset = @name = Robot.random_names.next
  
  def self.forget = random_names.rewind
  def self.random_names = @@robot_names ||= ('AA000'..'ZZ999').to_a.shuffle.to_enum
end

File: julia/robot-name/robot-name.jl
import Random: shuffle

names = [l1 * l2 * lpad(n, 3, '0') for l1 in 'A':'Z' for l2 in 'A':'Z' for n in 0:999] |> shuffle

mutable struct Robot
    name::String
end

Robot() = Robot(pop!(names))

reset!(instance::Robot) = instance.name = pop!(names)
name(instance::Robot) = instance.name
